### ISSUE TITLE
Incompatibility with optax.lbfgs and potentially other optimizers

### ISSUE THREAD

**Issue by jaosch** [Drenderer/klax#12]
**To reproduce:**
- Run `test_training` in `test_training.py`, but replace `optimizer.adam` with `optimizer.lbfgs`

The problem is probably caused, because some optimizers, not only need the gradient of the loss function, but also the loss value directly see [optax.lbrgs documentsion](https://optax.readthedocs.io/en/latest/api/optimizers.html#optax.lbfgs)

**TODO:**

- [ ]  Fix training.py to allow for different optax optimizers to be used
- [ ]  Parameterize `test_training` to test multiple optimizers at once.

**Comment by Drenderer** [comment]
Issue should be fixed now. However, some `optax` line-search optimizers are not currently implemented in the most efficient way.
